utils/template_renderer.py

```diff
@@ -0,0 +1,104 @@
+                self.lookup_dirs = [lookup_dirs]
+            else:
+                self.lookup_dirs = list(lookup_dirs)
+
+    def _find_absolute_path(self, path):
+        if os.path.isabs(path):
+            return path
+        for directory in self.lookup_dirs:
+            potential_location = pathlib.Path(directory / path)
+            if potential_location.exists():
+                return str(potential_location.absolute())
+        return path
+
+    def get_source(self, environment, template):
+        template = self._find_absolute_path(template)
+        return super().get_source(environment, template)
+
+
+class Renderer(object):
+    TEMPLATE_NAME = ""
+
+    def __init__(self, product, build_dir):
+        self.project_directory = pathlib.Path(os.path.dirname(__file__)).parent.resolve()
+
+        self.product = product
+        self.env_yaml = self.get_env_yaml(build_dir)
+
+        self.built_content_path = pathlib.Path(
+            "{build_dir}/{product}".format(build_dir=build_dir, product=product))
+
+        self.template_data = dict()
+
+    def get_env_yaml(self, build_dir):
+        product_yaml = self.project_directory / self.product / "product.yml"
+        build_config_yaml = pathlib.Path(build_dir) / "build_config.yml"
+        if not (product_yaml.exists() and build_config_yaml.exists()):
+            msg = (
+                "No product yaml and/or build config found in "
+                "'{product_yaml}' and/or '{build_config_yaml}', respectively, please make sure "
+                "that you got the product right, and that it is built."
+                .format(product_yaml=product_yaml, build_config_yaml=build_config_yaml)
+            )
+            raise ValueError(msg)
+        env_yaml = ssg.yaml.open_environment(build_config_yaml, product_yaml)
+        return env_yaml
+
+    def _set_rule_relative_definition_location(self, rule):
+        rule.relative_definition_location = (
+            pathlib.PurePath(rule.definition_location)
+            .relative_to(self.project_directory))
+
+    def get_result(self):
+        subst_dict = self.template_data.copy()
+        subst_dict.update(self.env_yaml)
+        html_jinja_template = os.path.join(os.path.dirname(__file__), self.TEMPLATE_NAME)
+        ssg.jinja._get_jinja_environment.env.loader = FlexibleLoader([self.project_directory / "utils"])
+        return ssg.jinja.process_file(html_jinja_template, subst_dict)
+
+    def output_results(self, args):
+        if "title" not in self.template_data:
+            self.template_data["title"] = args.title
+        result = self.get_result()
+        if not args.output:
+            print(result)
+        else:
+            with open(args.output, "w") as outfile:
+                outfile.write(result)
+
+    @staticmethod
+    def create_parser(description):
+        parser = argparse.ArgumentParser(description=description)
+        parser.add_argument(
+            "product", help="The product to be built")
```

Comment:
The missing dashes cause a problem that product is a positional argument when it should be optional.

```suggestion
            "--product", help="The product to be built")
```

Reply:
Why? The product is required, so why should it be an optional argument?

Reply:
Because you already have another positional argument and having 2 will be confusing

Reply:
I disagree, it's one product followed by whatever, in this case, that would be a set of one or more rules. The help message is rendered correctly, and what you propose is a required optional argument, which is [somehow against established conventions](https://docs.python.org/3/library/argparse.html#required) (see the box with the note).